source/particle/world.cc

```diff
@@ -501,6 +501,53 @@ namespace aspect
     }
 
     template <int dim>
+    bool
+    World<dim>::particle_is_in_cell(const Particle<dim> &particle,
+                                    const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const
+    {
+      try
+        {
+          const Point<dim> p_unit = this->get_mapping().transform_real_to_unit_cell(cell, particle.get_location());
+          if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+            return true;
+        }
+      catch (...)
+        {}
+      return false;
+    }
+
+    template <int dim>
+    std::map<double, typename parallel::distributed::Triangulation<dim>::active_cell_iterator>
+    World<dim>::neighbor_cells_to_search(const Particle<dim> &particle,
+                                         const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const
+    {
+      std::map<double,typename parallel::distributed::Triangulation<dim>::active_cell_iterator> neighbor_cells;
+      for (unsigned int face_no=0; face_no<GeometryInfo<dim>::faces_per_cell; ++face_no)
+        if (cell->at_boundary(face_no) == false)
+          {
+            if (cell->neighbor(face_no)->active())
+              {
+                const double center_distance = (particle.get_location() - cell->face(face_no)->center()).norm();
+                neighbor_cells.insert(std::make_pair(center_distance,cell->neighbor(face_no)));
+              }
+            else
+              // If the neighbor cell is not active (i.e it has
+              // children) cycle through them. If the children have
+              // children we can ignore those, since they can not
+              // be connected to our original cell, because level
+              // differences across faces are limited to 1.
+              for (unsigned int child_index=0; child_index<cell->neighbor(face_no)->n_children(); ++child_index)
+                if (cell->neighbor(face_no)->child(child_index)->active())
```

Comment:
I don't know if it's worth it, but you could use `cell->neighbor_child_on_subface` to only get the immediate neighbors.